### PR TITLE
Multiple load allocate

### PR DESCRIPTION
**Pull request by ViljoenFritz**
Summary of this change:

added feature to allocate pallets to multiple loads from text field 

CHANGELOG entry:

### Added
- .Parse load string
- allow loads to be deleted when allocated
### Changed
- rearrange loads ETD->ETA ATA->ATD
- updated factories
### Fixed
- .removed wip pallets check from interactor they are checked in check_pallets

How to test this code via the UI:

loads => manage => allocate => paste list of pallets in text field

### CODE REVIEW

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: helpers/utility_functions.rb [hunk 1]
```diff
@@ -291,6 +291,20 @@ def xml_file?(file_path)
...
+  def parse_string_to_array_of_hashes(in_string)
+    output = []
+    in_string.split(/\n/).map(&:strip).reject(&:empty?).each do |row|
+      row_hash = {}
+      row.split(/\t|,| /).each_with_index { |v, k| row_hash[k] = v }
+      output << row_hash
+    end
+    output
+  end
+
```
[the remark this answers is not in the document]
Comment (jcsjcs) [reply helpers/utility_functions.rb:307]: Yes, I think tabs will be a valuable addition.

[the remark this answers is not in the document]
Comment (ViljoenFritz) [reply helpers/utility_functions.rb]: applied changes

File: helpers/utility_functions.rb [hunk 2]
```diff
@@ -291,26 +291,12 @@ def xml_file?(file_path)
...
-    output = []
-    in_string.split(/\n/).map(&:strip).reject(&:empty?).each do |row|
-      row_hash = {}
-      row.split(/\t|,| /).each_with_index { |v, k| row_hash[k] = v }
-      output << row_hash
-    end
-    output
-  end
-
   # Takes a string and returns an array from text split on commas and new lines.
```
Comment (jcsjcs) [inline helpers/utility_functions.rb:294]: Amend the description to include "tabs".

File: lib/finished_goods/test/services/test_parse_load_string.rb [hunk 3]
```diff
@@ -43,7 +42,7 @@ def test_string_parse_success
       pallet_number3 = Faker::Number.number(digits: 9).to_s
 
       in_string = "1,#{pallet_number1}\n
-                   2 #{pallet_number2}\n
+                   2,#{pallet_number2}\n
                    3\t#{pallet_number3}"
 
       res = ParseLoadString.call(in_string)
```
Comment (jcsjcs) [inline lib/finished_goods/test/services/test_parse_load_string.rb:48]: These tests must also pass if there is extra empty space between delimiters:
`"    1  ,  #{pallet_number}  \t   1  \t  #{pallet_number} \n  "`

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: lib/finished_goods/repositories/load_repo.rb [hunk 4]
```diff
@@ -131,6 +137,9 @@ def allocate_pallets(load_id, pallet_numbers, user)
       # updates load status allocated
       update(:loads, load_id, allocated: true, allocated_at: Time.now)
       log_status(:loads, load_id, 'ALLOCATED', user_name: user.user_name)
+
+      FinishedGoodsApp::ProcessOrderLines.call(@user, load_id: load_id)
+      FinishedGoodsApp::Job::CalculateExtendedFgCodesFromSeqs.enqueue(pallet_ids) if AppConst::CR_FG.lookup_extended_fg_code?
```
Comment (jcsjcs) [inline lib/finished_goods/repositories/load_repo.rb:142]: Why is the calculation of extended fg called here?

File: lib/finished_goods/services/parse_load_string.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    attr_reader :parse_string
+
+    def initialize(parse_string)
+      @parse_string = parse_string
+    end
+
+    def call
+      output = parse_load_string
+
+      check_non_numeric(output)
```
Comment (jcsjcs) [inline lib/finished_goods/services/parse_load_string.rb:14]: Is this necessary if you have called `valid_pallet_number?` already?

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: lib/finished_goods/services/parse_load_string.rb (hunk 5 above)
Comment (ViljoenFritz) [reply lib/finished_goods/services/parse_load_string.rb:14]: MesscadaApp::ScannedPalletNumber doesnt check for non numeric

File: lib/finished_goods/test/services/test_parse_load_string.rb (hunk 3 above)
Comment (ViljoenFritz) [reply lib/finished_goods/test/services/test_parse_load_string.rb:48]: ill add in some spaces

File: lib/finished_goods/repositories/load_repo.rb (hunk 4 above)
Comment (ViljoenFritz) [reply lib/finished_goods/repositories/load_repo.rb:142]: https://github.com/NoSoft-SA/nspack/pull/334/commits/b2c1454b6caf3521a66b9a360f6c645c87d595d6
it was clogging up the interactor

File: helpers/utility_functions.rb (hunk 2 above)
Comment (ViljoenFritz) [reply helpers/utility_functions.rb]: done

File: lib/finished_goods/services/parse_load_string.rb (hunk 5 above)
Comment (ViljoenFritz) [reply lib/finished_goods/services/parse_load_string.rb:14]: ill rename valid_pallet_number to possable_pallet_number

**Pull request merged**